### PR TITLE
Back/signin

### PR DESCRIPTION
**Pull request by yimialmonte**
This is backend implementation for SignIn/Login functionality

### CODE REVIEW

File: components/users/user.js [hunk 1]
```diff
@@ -1,32 +1,73 @@
...
-  if(!user.isModified('password')) return
+  if (!user.isModified('password')) return
 
   user.password = await bcryptjs.hash(user.password, 8)
   next()
 })
 
+userSchema.methods.generateToken = async function () {
+  const user = this
+  const secretKey = process.env.SECRET_KEY || 'secret'
```
Comment (cesar18pena) [inline components/users/user.js:41]: We have a configuration file for `environment variable` let add this line there

File: components/users/user.js [hunk 2]
```diff
@@ -1,32 +1,73 @@
...
+
+  const errorMessage = 'Unable to login';
+
+  if(!user) throw new Error(errorMessage)
+
+  const passwordMatch = await bcryptjs.compare(password, user.password);
+  if(!passwordMatch) throw new Error(errorMessage)
+
+  return user
+}
```
Comment (cesar18pena) [inline components/users/user.js:60]: This method is called `findUser` but I see we are passing email and password and doing more things that just find. Let keep this simple, a function to receive an email and check is the user exists or not

File: components/users/user.js [hunk 3]
```diff
@@ -1,32 +1,73 @@
...
+
+userSchema.methods.toJSON = function() {
+  const user = this;
+  const userObject = user.toObject()
+  delete userObject.password
+  delete userObject.tokens
+  delete userObject._id
+  return userObject
+}
```
Comment (cesar18pena) [inline components/users/user.js:69]: Can you explain the use of this method? we are just deleting most of the properties

File: components/users/user.js (hunk 1 above)
Comment (yimialmonte) [reply components/users/user.js]: Fixed

File: components/users/user.js (hunk 2 above)
Comment (yimialmonte) [reply components/users/user.js]: I changed the name of the method, findUserByCredencials more clear name about what is really this method do

File: components/users/user.js (hunk 3 above)
Comment (yimialmonte) [reply components/users/user.js]: I simplified this method, the idea is to select the user properties that we want to send, which is like a mapper.

File: components/users/user.js [hunk 4]
```diff
@@ -1,32 +1,68 @@
...
-  if(!user.isModified('password')) return
+  if (!user.isModified('password')) return
 
   user.password = await bcryptjs.hash(user.password, 8)
   next()
 })
 
+userSchema.methods.generateToken = async function () {
+  const user = this
+  const secretKey = process.env.SECRET_KEY
```
Comment (cesar18pena) [inline components/users/user.js:41]: Add this env to our environment file and import it here to use

**Pull request closed without merging**